File: Controllers/UsersController.cs
using DMS.Models;
using DocumentFormat.OpenXml.Bibliography;
using DocumentFormat.OpenXml.Wordprocessing;
using Microsoft.AspNetCore.Http.HttpResults;
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json;
using System;
using System.Diagnostics;

namespace DMS.Controllers
{
    public class UsersController : Controller
    {
        public const string SessionKeyName = "username";
        private readonly ILogger<UsersController> _logger;
        private readonly IHttpContextAccessor contxt;

        private DMS_DbContext dataContext { get; set; }

        public UsersController(ILogger<UsersController> logger, DMS_DbContext _dataContext, IHttpContextAccessor httpContextAccessor)
        {
            _logger = logger;
            dataContext = _dataContext;
            contxt = httpContextAccessor;
        }

        public IActionResult Create()
        {
            ViewData["userAccessList"] = dataContext.UserAccess.ToList();
            TempData["username"] = HttpContext.Session.GetString(SessionKeyName);
            return View();
        }
        public IActionResult Index()
        {
            var finalListJson = TempData["FinalList"] as string;

            if (finalListJson != null)
            {
                var finalList = JsonConvert.DeserializeObject<List<UsersModel>>(finalListJson);

                foreach (var user in finalList)
                {
                    user.UserAccess = dataContext.UserAccess
                        .FirstOrDefault(ua => ua.ID == user.UserAccessID)?.Description;
                }

                return View(finalList);
            }

            var UsersList = (from user in dataContext.Users
                             join access in dataContext.UserAccess on user.UserAccessID equals access.ID into accessGroup
                             from access in accessGroup.DefaultIfEmpty()
                             where user.Active
                             select new UsersModel
                             {
                                 ID = user.ID,
                                 EmployeeID = user.EmployeeID,
                                 UserName = user.UserName,
                                 Password = user.Password,
                                 UserAccessID = user.UserAccessID,
                                 UserAccess = access != null ? access.Description : null,
                                 Active = user.Active,
                                 CreatedBy = user.CreatedBy,
                                 CreatedDate = user.CreatedDate
                             }).ToList();

            TempData["username"] = HttpContext.Session.GetString(SessionKeyName);
            return View(UsersList);
        }


        public IActionResult Delete(long id)
        {
            var obj = dataContext.Users.Find(id);
            if (obj == null)
                return NotFound();

            obj.Active = false;
            dataContext.SaveChanges();

            return RedirectToAction("Index");
        }


        [HttpPost]
        [ValidateAntiForgeryToken]
        public IActionResult Edit(UsersModel Users, string password, string cpassword)
        {
            var employeeID = (string)TempData["EmployeeID"];
            var userName = (string)TempData["UsersUserName"];
            var userAccessID = (string)TempData["UserAccessID"];
            var userPassword = (string)TempData["Password"];

            // Validate if the name and age exist in the database
            bool userNameExist = dataContext.Users.Any(p => p.UserName == Users.UserName && p.UserName != userName);

            var active = (bool)TempData["Active"];
            var CreatedBy = (string)TempData["CreatedBy"];
            var CreatedDate = (DateTime)TempData["CreatedDate"];

            if (userNameExist)
            {
                ViewBag.UserNameAlert = "<span class='text-danger'>This User Name already Exist</span>";
                LoadUserTemp(employeeID, userName, userAccessID, userPassword, active, CreatedBy, CreatedDate);
                return View(Users);
            }
            if (Users.UserName == null)
            {
                ViewBag.UserNameAlert = "<span class='text-danger'>Please Input a User Name</span>";
                LoadUserTemp(employeeID, userName, userAccessID, userPassword, active, CreatedBy, CreatedDate);
                return View(Users);
            }

            if (!string.IsNullOrEmpty(password))
            {
                if (string.IsNullOrEmpty(cpassword))
                {
                     ViewBag.PasswordAlert = "<span class='text-danger'>Please confirm your password</span>";
                     LoadUserTemp(employeeID, userName, userAccessID, userPassword, active, CreatedBy, CreatedDate);
                    return View(Users);
                }

                if (password != cpassword)
                {
                    ViewBag.PasswordAlert = "<span class='text-danger'>Passwords do not match</span>";
                    LoadUserTemp(employeeID, userName, userAccessID, userPassword, active, CreatedBy, CreatedDate);
                    return View(Users);
                }

                Users.Password = GlobalMethods.Encrypt(password);
            }
            else
            {
                Users.Password = userPassword;
            }



            Users.EmployeeID = int.Parse(employeeID);
            Users.CreatedDate = CreatedDate;
            Users.CreatedBy = CreatedBy;
            Users.Active = active;
            dataContext.Users.Update(Users);
            dataContext.SaveChanges();
            return RedirectToAction("Index");
        }


        [HttpPost]
        [ValidateAntiForgeryToken]
        public IActionResult Create(UsersModel Users, string password, string cpassword)
        {
            bool userNameExist = dataContext.Users.Any(p => p.UserName == Users.UserName);

            if (userNameExist)
            {
                ViewBag.UserNameAlert = "<span class='text-danger'>This User Name already Exist</span>";
                ViewData["userAccessList"] = dataContext.UserAccess.ToList();
                return View(Users);
            }
            if (Users.UserName == null)
            {
                ViewBag.UserNameAlert = "<span class='text-danger'>Please Input a User Name</span>";
                ViewData["userAccessList"] = dataContext.UserAccess.ToList();
                return View(Users);
            }

            if (!string.IsNullOrEmpty(password))
            {
                if (string.IsNullOrEmpty(cpassword))
                {
                    ViewBag.PasswordAlert = "<span class='text-danger'>Please confirm your password</span>";
                    ViewData["userAccessList"] = dataContext.UserAccess.ToList();
                    return View(Users);
                }

                if (password != cpassword)
                {
                    ViewBag.PasswordAlert = "<span class='text-danger'>Passwords do not match</span>";
                    ViewData["userAccessList"] = dataContext.UserAccess.ToList();
                    return View(Users);
                }
                Users.Password = GlobalMethods.Encrypt(password);
            }
            else
            {
                ViewBag.PasswordAlert = "<span class='text-danger'>Please Input a Password</span>";
                ViewData["userAccessList"] = dataContext.UserAccess.ToList();
                return View(Users);
            }

            Users.EmployeeID = 0;
            Users.CreatedDate = DateTime.Now;
            Users.Active = true;
            Users.CreatedBy = HttpContext.Session.GetString(SessionKeyName);
            dataContext.Users.Add(Users);
            dataContext.SaveChanges();
            return RedirectToAction("Index");
        }












        public void LoadUserTemp(string employeeID, string userName, string userAccessID, string userPassword , bool active, string CreatedBy, DateTime CreatedDate)
        {

            ViewData["userAccessList"] = dataContext.UserAccess.ToList();

            TempData["EmployeeID"] = employeeID;
            TempData["UserAccessID"] = userAccessID;
            TempData["UsersUserName"] = userName;
            TempData["Active"] = active;
            TempData["Password"] = userPassword;
            TempData["CreatedBy"] = CreatedBy;
            TempData["CreatedDate"] = CreatedDate;
        }

        public IActionResult Edit(long id)
        {
            if (id == null || id == 0)
                return NotFound();

            var user = dataContext.Users.Find(id);

            if (user == null)
                return NotFound();

            var userAccess = dataContext.UserAccess
        .FirstOrDefault(ua => ua.ID == user.UserAccessID);

            if (userAccess != null)
            {
                user.UserAccess = userAccess.Description;
            }

            TempData["EmployeeID"] = user.EmployeeID?.ToString();
            TempData["UserAccessID"] = user.UserAccessID?.ToString();
            TempData["UsersUserName"] = user.UserName;
            TempData["Password"] = user.Password;
            TempData["Active"] = user.Active;
            TempData["CreatedBy"] = user.CreatedBy;
            TempData["CreatedDate"] = user.CreatedDate;

            ViewData["userAccessList"] = dataContext.UserAccess.ToList();

            return View(user);
        }




        [ResponseCache(Duration = 0, Location = ResponseCacheLocation.None, NoStore = true)]
        public IActionResult Error()
        {
            return View(new ErrorViewModel { RequestId = Activity.Current?.Id ?? HttpContext.TraceIdentifier });
        }

    }
}

File: Controllers/UsersController.js
const express = require('express');
const bcrypt = require('bcrypt');
const { body, validationResult } = require('express-validator');
const db=require('../config/database'); // Adjust the path as necessary
const Users = db.Users
const UserAccess=db.UserAccess
const router = express.Router();
const requireAuth = require('../middleware/requireAuth'); 
// GET /create
router.get('/create', async (req, res) => {
    try {
        const userAccessList = await UserAccess.findAll();
        res.render('users/create', {
            userAccessList,
            username: req.session.username,
            errors: {},
            user: {}
        });
    } catch (error) {
        console.error('Error in create GET:', error);
        res.status(500).render('error', { error });
    }
});

// GET /
router.get('/',requireAuth, async (req, res) => {
    try {
        console.log("=== Route Handler Started ===");
    console.log("Request URL:", req.originalUrl);
    console.log("Request Method:", req.method);
        console.log("req.user====>",req.user)
        let usersList;

        // Check if there's filtered data in session
        if (req?.user?.finalList) {
            usersList = req?.user?.finalList;
            
            // Add UserAccess descriptions
            for (let user of usersList) {
                const userAccess = await UserAccess.findByPk(user.userAccessID);
                user.userAccess = userAccess ? userAccess.description : null;
            }
            
            // Clear the session data after use
            delete req.user.finalList;
        } else {
            // Fetch all active users with their access levels
            usersList = await Users.findAll({
                where: { Active: true },

                 include: [{
                    model: UserAccess,
                    include:[
                        {
                            model:db.ModuleAccess,
                            as:"moduleAccess"
                        }
                    ],
                    as: 'accessList'
                }]
                // include: [{
                //     model: UserAccess,
                //     as: 'userAccess',
                //     include: [{
                //         model: db.ModuleAccess, // Assuming you have a ModuleAccess model
                //         as: 'moduleAccess',
                //         include: [{
                //             model: db.Module, // Assuming you have a Module model   
                //             as: 'module',
                //         }]
                //     }]
                // }],
                
            });
            // console.log("usersList",usersList)
            // // Transform to match the expected format
            // usersList = usersList.map(user => ({
            //     id: user.ID,
            //     employeeID: user.EmployeeID,
            //     userName: user.UserName,
            //     password: user.Password,
            //     userAccessID: user.UserAccessID,
            //     userAccess: user.userAccess ? user.userAccess.description : null,
            //     active: user.active,
            //     createdBy: user.createdBy,
            //     createdDate: user.createdDate
            // }));
        }
        console.log("req.user",req.user)
        res.json( {
            users: usersList,
            username: req.user.username
        });
    } catch (error) {
        console.error('Error in index:', error);
        res.status(500).json({
            status:false
        });
    }
});




// router.get('/', requireAuth, async (req, res) => {
//     console.log("=== Route Handler Started ===");
//     console.log("Request URL:", req.originalUrl);
//     console.log("Request Method:", req.method);
    
//     try {
//         console.log("Checking req.user...");
//         console.log("req.user exists?", !!req.user);
//         console.log("req.user:", req.user);
//         console.log("typeof req.user:", typeof req.user);
        
//         // Check all properties on req object
//         console.log("All req properties containing 'user':");
//         Object.keys(req).forEach(key => {
//             if (key.toLowerCase().includes('user')) {
//                 console.log(`req.${key}:`, req[key]);
//             }
//         });
        
//         // Check if user data exists anywhere else
//         console.log("req.headers:", req.headers);
        
//         if (!req.user) {
//             console.log("req.user is null/undefined - this is the problem!");
//             return res.status(401).json({ error: 'User not authenticated' });
//         }
        
//         console.log("Success! req.user is available:", req.user);
//         res.json({
//             message: 'Success',
//             user: req.user
//         });
        
//     } catch (error) {
//         console.error('Route error:', error);
//         res.status(500).json({ error: 'Internal server error' });
//     }
// });


// GET /delete/:id (soft-delete only active users)
router.get('/delete/:id', async (req, res) => {
    try {
        const id = parseInt(req.params.id);
        // Only find active users to avoid re-deleting inactive ones
        const user = await Users.findOne({
            where:{
                ID:id,
                Active: true
            }
        });
        console.log("user",user)
        if (!user) {
            return res.json({
                status:false,
                message:"User not found"
            })
        }

        // Soft delete - set active to false
        await Users.update({ Active: false },{where:{ID:id}});
        
        res.json({
            message:"deleted"
        });
    } catch (error) {
        console.error('Error in delete:', error);
        res.status(500).render('error', { error });
    }
});

router.delete('/delete/:id', async (req, res) => {
    try {
        const id = parseInt(req.params.id);

        // Soft delete instead of hard destroy: set Active = false
        const [updated] = await Users.update(
            { Active: false },
            { where: { ID: id } }
        );

        if (updated === 0) {
            return res.status(404).json({ status: false, message: 'User not found' });
        }

        res.json({
            status: true,
            message: "User deleted"
        });
    } catch (error) {
        console.error('Error in delete:', error);
        res.status(500).json({ error: error.message });
    }
});

// GET /edit/:id — only allow editing active users
router.get('/edit/:id', async (req, res) => {
    try {
        const id = parseInt(req.params.id);
        
        if (!id) {
            return res.status(404).render('error', { 
                error: { message: 'User not found' } 
            });
        }

        // Only fetch user if active
        const user = await Users.findOne({ where: { ID: id, Active: true } });
        if (!user) {
            return res.status(404).render('error', { 
                error: { message: 'User not found' } 
            });
        }

        const userAccess = await UserAccess.findByPk(user.userAccessID);
        if (userAccess) {
            user.userAccess = userAccess.description;
        }

        // Store original data in session for comparison during update
        req.session.editUserData = {
            employeeID: user.employeeID,
            originalUserName: user.userName,
            userAccessID: user.userAccessID,
            password: user.password,
            active: user.active,
            createdBy: user.createdBy,
            createdDate: user.createdDate
        };

        const userAccessList = await UserAccess.findAll();

        res.json({
            user: user.toJSON(),
            userAccessList,
            errors: {},
            username: req.session.username
        });
    } catch (error) {
        console.error('Error in edit GET:', error);
        res.status(500).render('error', { error });
    }
});

// POST /create
router.post('/create', [
    body('userName').notEmpty().withMessage('Please Input a User Name'),
    body('password').notEmpty().withMessage('Please Input a Password'),
    body('cpassword').custom((value, { req }) => {
        if (value !== req.body.password) {
            throw new Error('Passwords do not match');
        }
        return true;
    })
], async (req, res) => {
    try {
        const errors = validationResult(req);
        const { userName, password, userAccessID } = req.body;

        let validationErrors = {};
        
        if (!errors.isEmpty()) {
            errors.array().forEach(error => {
                validationErrors[error.param] = error.msg;
            });
        }

        // Check if username already exists
        const existingUser = await Users.findOne({ where: { userName } });
        if (existingUser) {
            validationErrors.userName = 'This User Name already exists';
        }

        if (Object.keys(validationErrors).length > 0) {
            const userAccessList = await UserAccess.findAll();
            return res.render('users/create', {
                userAccessList,
                errors: validationErrors,
                user: req.body,
                username: req.session.username
            });
        }

        // Hash the password
        const hashedPassword = await encryptPassword(password);

        // Create new user
        await Users.create({
            employeeID: 0,
            userName,
            password: hashedPassword,
            userAccessID: parseInt(userAccessID),
            active: true,
            createdBy: req.session.username,
            createdDate: new Date()
        });

        res.json({
            message:"success"
        });

    } catch (error) {
        console.error('Error in create POST:', error);
        res.status(500).render('error', { error });
    }
});

// POST /edit
router.post('/edit', [
    body('userName').notEmpty().withMessage('Please Input a User Name'),
    body('cpassword').custom((value, { req }) => {
        if (req.body.password && value !== req.body.password) {
            throw new Error('Passwords do not match');
        }
        if (req.body.password && !value) {
            throw new Error('Please confirm your password');
        }
        return true;
    })
], async (req, res) => {
    try {
        const errors = validationResult(req);
        const { id, userName, password, userAccessArray } = req.body;
        // const editData = req.session.editUserData;

        // if (!editData) {
        //     return res.redirect('/users');
        // }

        let validationErrors = {};
        
        if (!errors.isEmpty()) {
            errors.array().forEach(error => {
                validationErrors[error.param] = error.msg;
            });
        }

        // Check if username already exists (excluding current user)
        const existingUser = await Users.findOne({ 
            where: { 
                userName,
                id: { [require('sequelize').Op.ne]: parseInt(id) }
            }
        });
        
        if (existingUser) {
            validationErrors.userName = 'This User Name already exists';
        }

        let parseArray;

        try {
            parseArray = JSON.parse(userAccessArray);
        } catch {
            parseArray = userAccessArray;
        }

        console.log("userAccessArray", userAccessArray);

        // 🔴 Delete all previous accesses for this user
        await db.UserUserAccess.destroy({
            where: { UserID: id }
        });

        // 🟢 Now create new ones
        for (let i = 0; i < parseArray.length; i++) {
            await db.UserUserAccess.create({
                UserID: id,
                UserAccessID: parseArray[i]
            });
        }


        // Prepare update data
        const updateData = {
            UserName: userName,
            userAccessArray: userAccessArray || [], 
            
           
        };

        // Update password if provided
        if (password && password.trim() !== '') {
            updateData.Password = await encryptPassword(password);
        }
        console.log("updateData",updateData)
        await Users.update(updateData, {
            where: { ID: parseInt(id) }
        });
        
        // Clear session data
        return res.json({
            status:true,
            message: "User updated successfully",
           
        });

    } catch (error) {
        console.error('Error in edit POST:', error);
        res.status(500).json({ error });
    }
});

// Helper function to encrypt password
async function encryptPassword(password) {
    const saltRounds = 10;
    return await bcrypt.hash(password, saltRounds);
}

// Helper function to verify password
async function verifyPassword(password, hashedPassword) {
    return await bcrypt.compare(password, hashedPassword);
}

module.exports = router;